### PR TITLE
Fix enough code to get make test_prep working with -Dusedefaultstrict

### PR DESCRIPTION
**Pull request by toddr**
This change fixes enough code that it's possible to run make test_prep when perl is compiled with -Dusedefaultstrict.

There are 2 caveats to this:

 - Does not address XSLoader/DynaLoader already submitted in another PR (#18221)
 - Does not address cpan/Pod-Usage or  cpan/Text-Tabs which continue to be outstanding upstream.

### CODE REVIEW

File: lib/utf8.pm [hunk 1]
```diff
@@ -1,15 +1,19 @@
 package utf8;
 
-$utf8::hint_bits = 0x00800000;
```
Comment (toddr) [inline lib/utf8.pm:3]: I'm aware this might have unexpected side effects but I wasn't able to find any. I'm happy to revert this if there's a good reason for not allowing strict/warnings here. Is this a circular deps issue maybe?

File: lib/overload.pm [hunk 2]
```diff
@@ -1239,7 +1243,7 @@ Put this in F<symbolic.pm> in your Perl library directory:
 
 This module is very unusual as overloaded modules go: it does not
 provide any usual overloaded operators, instead it provides an
-implementation for C<L</nomethod>>.  In this example the C<nomethod>
+implementation for L</C<nomethod>>.  In this example the C<nomethod>
```
Comment (khwilliamson) [inline lib/overload.pm]: Why this change?  It's my experience that the original method looks better for internal links; the changed-to form for external links.  Sadly.

File: lib/utf8.pm (hunk 1 above)
Comment (khwilliamson) [reply lib/utf8.pm:3]: I don't know if this has side effects or not.

BTW I hate that we don't derive the bit number.

Comment (khwilliamson) [reply lib/utf8.pm:3]: So, apply it and see what breaks on cpan, hopefully nothing

File: lib/overload.pm [hunk 3]
```diff
@@ -1,8 +1,12 @@
 package overload;
 
-our $VERSION = '1.32';
+use strict;
+no strict 'vars'; # Needed for assignment to typeglobs and symbol table
```
Comment (Grinnz) [inline lib/overload.pm]: This comment is not accurate; assigning to typeglobs and symbol table entries works fine under strict 'vars'.

File: lib/overload.pm (hunk 2 above)
Comment (toddr) [reply lib/overload.pm]: That's an oversight. I have no opinion on POD style personally. I will remove it.

File: lib/overload.pm (hunk 3 above)
Comment (toddr) [reply lib/overload.pm]: What would you say instead?

File: lib/utf8.pm (hunk 1 above)
Comment (khwilliamson) [reply lib/utf8.pm:3]: I can't remember, but isn't this global the way it tells the core that 'use utf8' is in effect?  If so, this would break lots.

Comment (Grinnz) [reply lib/utf8.pm:3]: I don't know about possibility of circular reference, but applying strict and warnings to this module could be a significant performance hit due to how much it is used. IMO making the other changes and leaving out strict and warnings would be preferable in this specific case. By the way, bytes.pm is a "mirror" of this module so should be updated in the same way.

Comment (Grinnz) [reply lib/utf8.pm:3]: @khwilliamson This global is still being set a few lines below.

File: lib/overload.pm (hunk 3 above)
Comment (Grinnz) [reply lib/overload.pm]: That depends why it's needed here. Strict 'vars' requires you to declare a (scalar, array, or hash) variable with `my` or `our` before using it.

File: lib/utf8.pm (hunk 1 above)
Comment (toddr) [reply lib/utf8.pm:3]: I hear what you're saying. I hesitated for similar reasons in all honesty. 

 - require is almost no cost if the module is already loaded. This would only happen the first time utf8.pm is loaded anyway.
 - strict/warnings->import would also only happen once. Why would that be a performance hit?

File: lib/overload.pm (hunk 3 above)
Comment (toddr) [reply lib/overload.pm]: yep. nevermind. fixed.

File: lib/utf8.pm (hunk 1 above)
Comment (Grinnz) [reply lib/utf8.pm:3]: I'm not talking about the performance hit of activating strict/warnings but of having them active each time this module is imported or its AUTOLOAD is called. Though there are certainly bigger concerns at that point. I certainly don't think it will make a big difference but just questioning the tradeoff for this module in particular, maybe that's why they weren't set before, maybe not.

File: ext/XS-APItest/APItest_BS [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #
 # test that non-empty .bs files get executed (but only once)
-
+no strict 'vars';
```
Comment (Grinnz) [inline ext/XS-APItest/APItest_BS]: This doesn't seem like the right thing to do, shouldn't we be testing that these files continue to be executed in a non-strict context?

**Comment by toddr** [comment]
> So, apply it and see what breaks on cpan, hopefully nothing

I was horrified to find out how many inc/utf8.pm files there are on CPAN. Ignoring those leaves us with this:

```
$>git grep utf8::hint_bits|grep -v inc/utf8.pm
A/Acme-Emoticarp/lib/Acme/Emoticarp.pm:    $^H |= $utf8::hint_bits; ## no critic (Bangs::ProhibitBitwiseOperators)
E/Encode/encoding.pm:        require utf8;      # to fetch $utf8::hint_bits;
E/Encode/encoding.pm:        $^H |= $utf8::hint_bits;
F/Fennec/lib/Fennec/Util.pm:    my $with_utf8 = ( caller(0) )[8] & $utf8::hint_bits;
G/Game-Collisions/nytprof/utf8-pm-16-line.html:<tr><td class="h"><a name="3"></a>3</td><td class="c3">1</td><td class="c3"><span title="Avg 400ns">400ns</span></td><td></td><td></td><td class="s">$utf8::hint_bits = 0x00800000;</td></tr>
G/Game-Collisions/nytprof/utf8-pm-16-line.html:<tr><td class="h"><a name="8"></a>8</td><td class="c3">1</td><td class="c0"><span title="Avg 12µs">12µs</span></td><td></td><td></td><td class="s">    $^H |= $utf8::hint_bits;</td></tr>
G/Game-Collisions/nytprof/utf8-pm-16-line.html:<tr><td class="h"><a name="12"></a>12</td><td></td><td></td><td></td><td></td><td class="s">    $^H &amp;= ~$utf8::hint_bits;</td></tr>
b/bare/bare.pm:  my $legal_var = $utf8::hint_bits && $^H & $utf8::hint_bits
l/lambda/lib/lambda.pm:  $^H |= $utf8::hint_bits;
l/lambda/lib/lambda.pm:  $^H &= ~$utf8::hint_bits;
p/perl/cpan/Encode/encoding.pm:        require utf8;      # to fetch $utf8::hint_bits;
p/perl/cpan/Encode/encoding.pm:        $^H |= $utf8::hint_bits;
p/perl/lib/utf8.pm:$utf8::hint_bits = 0x00800000;
p/perl/lib/utf8.pm:    $^H |= $utf8::hint_bits;
p/perl/lib/utf8.pm:    $^H &= ~$utf8::hint_bits;
p/perl/pod/perldata.pod:          (?(?{ (caller(0))[8] & $utf8::hint_bits })
p/perl_mlb/os2/encoding.pm:	require utf8; # to fetch $utf8::hint_bits;
p/perl_mlb/os2/encoding.pm:	$^H |= $utf8::hint_bits;
p/perl_mlb/utf8.pm:$utf8::hint_bits = 0x00800000;
p/perl_mlb/utf8.pm:    $^H |= $utf8::hint_bits;
p/perl_mlb/utf8.pm:    $^H &= ~$utf8::hint_bits;
q/qbit/lib/qbit.pm:    $^H |= $utf8::hint_bits;

```

File: lib/Getopt/Std.pm [hunk 5]
```diff
@@ -119,6 +122,7 @@ sub getopt (;$$) {
 	        $$hash{$first} = $rest;
 	    }
 	    else {
+            no strict 'refs';
```
Comment (Grinnz) [inline lib/Getopt/Std.pm:125]: Indentation doesn't quite match the following lines.

Comment (toddr) [reply lib/Getopt/Std.pm:125]: Welcome to the already existing world of Mixed tabs/spaces. Can I put you down as a yes vote for doing an eventual mass perltidy of blead? 😛

**Comment by Grinnz** [comment]
> > So, apply it and see what breaks on cpan, hopefully nothing
> 
> I was horrified to find out how many inc/utf8.pm files there are on CPAN.

This shouldn't matter, `$utf8::hint_bits` is still being set with these changes

**Comment by toddr** [comment]
> > > So, apply it and see what breaks on cpan, hopefully nothing
> > 
> > 
> > I was horrified to find out how many inc/utf8.pm files there are on CPAN.
> 
> This shouldn't matter, `$utf8::hint_bits` is still being set with these changes

Yep. I'm just documenting what I found. Agreed. Looks fine to me.

File: lib/Getopt/Std.pm (hunk 5 above)
Comment (Grinnz) [reply lib/Getopt/Std.pm:125]: I wouldn't go as far as mass perltidy, but making indentation consistent as an isolated commit would be useful IMO

File: ext/XS-APItest/APItest_BS (hunk 4 above)
Comment (toddr) [reply ext/XS-APItest/APItest_BS]: The goal for this commit is to have a perl we can build with strict on everywhere by default. That means this code snippet starts with strict on and has to have it disabled. DynaLoader cannot influence the strictness of this section unless we suck in the BS and then do a quoted eval. but that's a topic for another pull request.

See also https://github.com/Perl/perl5/pull/18221#issuecomment-706311030

Comment (toddr) [reply ext/XS-APItest/APItest_BS]: I've removed this file from this commit and am moving BS changes over to #18221

File: lib/Getopt/Std.pm [hunk 6]
```diff
@@ -251,6 +257,7 @@ sub getopts ($;$) {
 		    $$hash{$first} = $rest;
 		}
 		else {
+            no strict 'refs';
```
Comment (atoomic) [inline lib/Getopt/Std.pm:260]: indentation looks weird, not sure if this is the original or your patch, but they are not matching (same with the next line)

File: lib/Symbol.pm [hunk 7]
```diff
@@ -132,6 +134,7 @@ sub qualify ($;$) {
 }
 
 sub qualify_to_ref ($;$) {
```
Comment (atoomic) [inline lib/Symbol.pm:136]: maybe we could consider to use function signature instead of prototype? (as part of a next PR)

File: lib/Symbol.t [hunk 8]
```diff
@@ -87,6 +90,7 @@ use Symbol qw(qualify qualify_to_ref);  # must import into this package too
 
 # tests for delete_package
 package main;
+no warnings 'once';
 $Transient::variable = 42;
```
Comment (atoomic) [inline lib/Symbol.t:94]: rather than disabling that warning in the global scope either
`$Transient::variable = $Transient::variable = 42;`
or 
`{ no warnings 'once'; $Transient::variable = 42; }`

File: lib/locale.pm [hunk 9]
```diff
@@ -70,7 +73,7 @@ sub import {
                               :numeric :monetary :time) );
         for (my $i = 0; $i < @_; $i++) {
```
Comment (atoomic) [inline lib/locale.pm:74]: not related to this change but maybe consider something like this `foreach my $arg ( @_ )`

File: lib/overload.pm [hunk 10]
```diff
@@ -26,7 +29,7 @@ my %ops_seen;
 sub nil {}
 
 sub OVERLOAD {
-  $package = shift;
+  my $package = shift;
   my %arg = @_;
```
Comment (atoomic) [inline lib/overload.pm:33]: `my ( $package, %args ) = @_;`

File: lib/utf8.pm [hunk 11]
```diff
@@ -1,15 +1,19 @@
...
+use strict;
+use warnings;
 
-our $VERSION = '1.23';
+our $hint_bits = 0x00800000;
```
Comment (atoomic) [inline lib/utf8.pm:6]: this is `HINT_UTF8` at minimum we probably want a test to confirm it's in sync with `perl.h`

File: lib/Symbol.t (hunk 8 above)
Comment (toddr) [reply lib/Symbol.t:94]: The test is complicated enough that I decided a more directed follow on fix would be a better choice for this.

File: lib/utf8.pm (hunk 11 above)
Comment (toddr) [reply lib/utf8.pm:6]: Right. This sounds like a separate case.

**Review by atoomic — APPROVED** [review]

File: lib/Symbol.pm (hunk 7 above)
Comment (Grinnz) [reply lib/Symbol.pm:136]: That's nonsense. Signatures and prototypes are unrelated features.
